Figgle.Generator/EmbedFontSourceGenerator.cs

```diff
@@ -0,0 +1,329 @@
+
+                    if (matchingExternalFont.FontDescriptionString is null)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(
+                            ErrorReadingExternalFontFileDiagnostic,
+                            embedFontInfo.Location ?? generationInfo.TargetType.Locations[0],
+                            embedFontInfo.FontName));
+                        continue;
+                    }
+
+                    fontDescription = matchingExternalFont.FontDescriptionString;
+                }
+
+                renderInfoBuilder.Add(new(
+                    embedFontInfo.MemberName!,
+                    embedFontInfo.FontName!,
+                    fontDescription));
+            }
+
+            context.AddSource(
+                $"{generationInfo.TargetType.Name}.g.cs",
+                RenderSource(generationInfo.TargetType, renderInfoBuilder.ToImmutable()));
+        });
+    }
+
+    private static bool IsValidTypeForGeneration(
+        SourceProductionContext context,
+        ITypeSymbol typeToGenerate)
+    {
+        // we don't need to check if the type is abstract because the C# compiler will error if
+        // a static class is abstract
+        if (typeToGenerate.IsStatic == false)
+        {
+            context.ReportDiagnostic(Diagnostic.Create(
+                TypeMustBeStaticDiagnostic,
+                typeToGenerate.Locations[0],
+                typeToGenerate.ToDisplayString(_fullyQualifiedFormat)));
+            return false;
+        }
+
+        if (typeToGenerate.ContainingType is not null)
+        {
+            context.ReportDiagnostic(Diagnostic.Create(
+                NestedTypeIsNotSupportedDiagnostic,
+                typeToGenerate.Locations[0],
+                typeToGenerate.ToDisplayString(_fullyQualifiedFormat)));
+            return false;
+        }
+
+        return true;
+    }
+
+    private string RenderSource(ITypeSymbol type, ImmutableArray<RenderSourceInfo> fontsToGenerate)
+    {
+        return $$"""
+            {{Header}}
+
+            using System;
+            using System.Collections.Concurrent;
+            using Figgle;
+
+            namespace {{type.ContainingNamespace.ToDisplayString(_fullyQualifiedFormat)}}
+            {
+                {{GetAccessibility(type.DeclaredAccessibility)}} static partial class {{type.Name}}
+                {
+                    private static readonly ConcurrentDictionary<string, FiggleFont> _fontByName = new(StringComparer.Ordinal);
+                    private static readonly StringPool _stringPool = new();
+                    {{RenderFonts(fontsToGenerate, indentationLevel: 2)}}
+                }
+            }
+            
+            """;
+
+        static string RenderFonts(ImmutableArray<RenderSourceInfo> renderInfos, int indentationLevel)
+        {
+            string indentation = new(' ', indentationLevel * 4);
+            var builder = new StringBuilder();
```

Comment:
Wonder if we can provide a decent guess at a capacity here. Or maybe just start with a capacity larger than the default of 16. Most fonts will be large. Maybe start with 4096 and avoid a bunch of temporary allocations as it grows.